The Word Game

danilomizin opened the pull request.

danilomizin committed:
Practice 9

danilomizin committed:
Practice 10

danilomizin committed:
Practice 12 and 13

danilomizin committed:
Practice 14

danilomizin committed:
add words_ps3.txt and initial ps3

danilomizin committed:
Game The Word Game

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    for i in range(num_vowels, n):
+        x = random.choice(CONSONANTS)
+        hand[x] = hand.get(x, 0) + 1
+    while True:
+        litter = random.choice(VOWELS)
+        if litter in hand.keys() and hand[litter] == 1:
+            hand["*"] = hand.pop(litter)
+            return hand
```
You can initially add wildcard to hand and reduce count of vowels in `for` loop.
It will be more effective and simple.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    returns: int >= 0
+    """
+    count = 0
+    for litter in word.lower():
```
By `litter` do you mean `letter`?

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    """
+    new_hand = hand.copy()
+    for litter in word.lower():
+        if litter in new_hand.keys() and new_hand[litter] > 0:
```
You can remove `.keys()`

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    returns: boolean
+    """
+    word = word.lower()
+    if "*" in word:
+        for vowel in VOWELS:
+            if word.replace("*", vowel) in word_list:
+                return True
+        else:
+            return False
```
There is a chance to use regular expressions.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+            return False
+    elif word in word_list:
+        for litter in word:
+            if (litter not in hand.keys()) or word.count(litter) > hand[litter]:
```
You can remove one pair of brackets. Firstly Python process `not`, then `and`, then `or` logic operators.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    if calculate_handlen(hand) != 0:
+        print("Current Hand:", end=" ")
+        display_hand(hand)
+        user_word = str(input("Enter word, or “!!” to indicate that you are finished:"))
```
Bare `input` takes input from console as string.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+        user_word = str(input("Enter word, or “!!” to indicate that you are finished:"))
+        if is_valid_word(user_word, hand, word_list):
+            count_points += get_word_score(user_word, calculate_handlen(hand))
+            print(user_word, "earned", str(get_word_score(user_word, calculate_handlen(hand))) + ".", " Total:",
```
To simplify `print`, use formatting.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+        elif user_word == "!!":
+            return count_points
+        else:
+            print("This is not valid word. Please choose another word", "\n")
```
There is no need in `\n` in the end, as print automatically move cursor to the next line in console.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+            return count_points
+        else:
+            print("This is not valid word. Please choose another word", "\n")
+        return play_hand(update_hand(hand, user_word), word_list, count_points)
```
It would be better to replace recursion with loop.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    Returns the score for a word. Assumes the word is a
+    valid word.
+
+    You may assume that the input word is always either a string of letters,
```
You can remove long and unnecessary parts of docs. to function.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    type_letter = ""
+    if letter in hand.keys():
+        new_hand = hand.copy()
+        if letter in VOWELS:
+            type_letter = VOWELS
+        elif letter in CONSONANTS:
+            type_letter = CONSONANTS
```
You can delete such conditions and make one instruction.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+    for i in range(num_vowels, n):
+        x = random.choice(CONSONANTS)
+        hand[x] = hand.get(x, 0) + 1
+    while True:
```
Such `while True:` condition is not a good practice.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,346 @@
+        print(" ")
+        count_of_hand = play_hand(hand, word_list, 0)
+        count_final_points += count_of_hand
+        print("Total score for this hand:", count_of_hand, "\n------")
```
Divide one print to several prints for better visual perception.

MaxShvtsv reviewed (changes requested):
Finished reviewing.

danilomizin committed:
Removed long and unnecessary parts of docs, changed while True to loop for and
in is_valid_word change loop for regular expressions and in play_hand change regular expressions for loop while

danilomizin commented:
> Finished reviewing.

Changed code to the your requirements.

danilomizin committed:
Merge Practice 9

danilomizin committed:
Merge Practice 10

danilomizin committed:
Merge Practice 12, 13

danilomizin committed:
Merge Practice 14

danilomizin committed:
Added regular expression

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,276 @@
+        x = random.choice(CONSONANTS)
+        hand[x] = hand.get(x, 0) + 1
+    hand["*"] = 1
+    for letter in hand:
+        if letter in VOWELS and hand[letter] == 1:
+            del hand[letter]
+            return hand
```
What if the count of specific vowel is not 1?
There is a chance that letters will be set like {'a': 3, 'b': 1, ...}

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,276 @@
+    """
+    new_hand = hand.copy()
+    for letter in word.lower():
+        if letter in new_hand and new_hand[letter] > 0:
```
Use `new_hand.get(letter, 0)` instead of this. There is a chance that the letter in word will not be in hand, and Python will throw an error.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,276 @@
+            return False
+    elif word in word_list:
+        for letter in word:
+            if letter not in hand or word.count(letter) > hand[letter]:
```
The same mistake. Fix this KeyError.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,276 @@
+    """
+    Allows the user to play the given hand, as follows:
+
+    * The hand is displayed.
```
Remove such long and unnecessary parts of docs.

MaxShvtsv commented on ps3.py:
```diff
@@ -0,0 +1,276 @@
+        display_hand(hand)
+        user_word = input("Enter word, or “!!” to indicate that you are finished:")
+        if is_valid_word(user_word, hand, word_list):
+            count_points += get_word_score(user_word, calculate_handlen(hand))
+            print(f"{user_word} earned {str(get_word_score(user_word, calculate_handlen(hand)))} ."
```
These instructions are repeating. It will be more simple to decrale new var with `get_word_score` value and pass it here.

MaxShvtsv reviewed (changes requested):
Finished reviewing.
